src/gsy_e/models/strategy/energy_parameters/heatpump/heat_pump.py

```diff
@@ -252,73 +394,56 @@ def _rotate_profiles(self, current_time_slot: Optional[DateTime] = None):
             self._heat_demand_Q_J.read_or_rotate_profiles()
 
     def _calc_energy_to_buy_maximum(self, time_slot: DateTime) -> float:
-        cop = self._state.heatpump.get_cop(time_slot)
-        if cop == 0:
-            return 0
-        max_energy_consumption_kWh = self._state.tanks.get_max_energy_consumption(cop, time_slot)
-        assert max_energy_consumption_kWh > -FLOATING_POINT_TOLERANCE
-        if max_energy_consumption_kWh > self._max_energy_consumption_kWh:
-            return self._max_energy_consumption_kWh
-        return max_energy_consumption_kWh
+        return self.combined_state.get_energy_to_buy_maximum_kWh(
+            time_slot, self._source_temp_C.profile[time_slot]
+        )
 
     def _calc_energy_to_buy_minimum(self, time_slot: DateTime) -> float:
-        cop = self._state.heatpump.get_cop(time_slot)
-        if cop == 0:
-            return 0
-        min_energy_consumption_kWh = self._state.tanks.get_min_energy_consumption(cop, time_slot)
-        if min_energy_consumption_kWh > self._max_energy_consumption_kWh:
-            return self._max_energy_consumption_kWh
-        return min_energy_consumption_kWh
+        return self.combined_state.get_energy_to_buy_minimum_kWh(
+            time_slot, self._source_temp_C.profile[time_slot]
+        )
+
+    def _update_last_time_slot_data(self, time_slot: DateTime):
+        last_time_slot = self.last_time_slot(time_slot)
+        if last_time_slot not in self._source_temp_C.profile:
+            return
+        self.combined_state.increase_tank_temp_from_bought_energy(
+            last_time_slot, self._bought_energy_kWh, self._source_temp_C.profile[time_slot]
+        )
+        self._bought_energy_kWh = 0.0
+        self._state.tanks.update_tanks_temperature(last_time_slot)
```

Comment:
AFAIS, this line is doing the same as line 410. Please correct me.

Reply:
This might have changed by the last commit, however this is not the same. `increase_tank_temp_from_bought_energy` only updates the `temp_increase_K` of the tanks, while the latter updates the `_storage_temp_C` by subtracting `increase_temp_K - decrease_temp_K`.  If you would prefer that, I can remove one of the 2 functions and merge the 2 operations in one. Let me know what you think!